src/frontend/src/helpers/websocket.ts

```diff
@@ -0,0 +1,391 @@
+    return true;
+  }
+
+  public isConnected(): boolean {
+    return this.connected;
+  }
+
+  public isOffline(): boolean {
+    return this.offlineMode === 2 || (!this.connected && this.offlineMode === 1);
+  }
+
+  public offlineEnabled(): boolean {
+    return this.offlineMode === 1 || this.offlineMode === 2;
+  }
+
+  public setOfflineModeSetting(setting: number): void {
+    var old = this.offlineMode;
+    if(setting === 0 || setting === 1 || setting === 2) {
+      this.offlineMode = setting;
+      // Set cookie, set expiry date to some date sufficiently in the future.
+      var expiryDate = new Date();
+      expiryDate.setFullYear(expiryDate.getFullYear() + 10);
+      /* TODO set cookie here
+      cookies.put(SEASHELL_OFFLINE_MODE_COOKIE, setting,
+        {secure: true, expires: expiryDate});
+      */
+    } else {
+      throw new WebsocketError("Invalid offline mode setting selected.", setting);
+    }
+    if(old === 2 && this.offlineMode !== old) {
+      // trigger reconnect and sync
+      if(!this.connected)
+        this.connect();
+      else {
+        this.invoke_cb('connected');
+        this.syncAll();
+      }
+    } else if(this.offlineMode === 2) {
+      this.invoke_cb('connected', this.offlineMode);
+    }
+  }
+
+  private make_offline_disabled(name: string) {
+    return async (...args: any[])=>{
+      if(!this.isOffline()) {
+        var res = await this.socket[name].apply(this.socket, args);
+        await this.syncAll();
+        return res;
+      } else {
+        this.rejectOffline(name);
+      }
+    };
+  }
+
+  private make_offline_noop(name: string) {
+    return async (...args: any[])=>{
+      if(!this.isOffline()) {
+        return await this.socket[name].apply(this.socket, args);
+      } else {
+        return true;
+      }
+    };
+  }
+
+  private make_offline_enabled(name: string, offlineWriteThrough?: boolean) {
+    var online_arity = this.socket[name].length;
+    var offline_arity = this.storage[name].length;
+    if(offline_arity != online_arity && offline_arity != online_arity+1) {
+      throw new WebsocketError("Offline and online arities differ.",
+        {offline: offline_arity, online:online_arity});
+    }
+    if(online_arity != offline_arity) {
+      console.log("Registering function %s which will use online result in write-through mode", name);
+    }
+    return async (...args: any[])=>{
+      if(args.length > online_arity) {
+        throw new WebsocketError("Too many arguments passed to "+name+".",
+          {received: args.length, expected: online_arity});
+      }
+      if(this.offlineEnabled()) {
+        var init = await this.storage.isInitialized();
+        if(!init && !this.isOffline()) {
+          await this.syncAll();
+        } else if(!init) {
+          throw new WebsocketError("Cannot call "+name+" in offline mode before an initial sync.");
+        }
+        var res = await this.storage[name].apply(this.storage, args);
+      }
```

Comment:
Can anyone tell me why this is needed?
```
+      if(this.offlineEnabled()) {
 +        var init = await this.storage.isInitialized();
 +        if(!init && !this.isOffline()) {
 +          await this.syncAll();
 +        } else if(!init) {
 +          throw new WebsocketError("Cannot call "+name+" in offline mode before an initial sync.");
 +        }
 +        var res = await this.storage[name].apply(this.storage, args);
 +      }
```
Does this mean if the offline mode is on "enabled", and storage is dirty, `this.syncAll()` is never going to be called here for any of those `offline_enabled` functions?